### PR TITLE
Check `-parallel-jobs` before use

### PR DESCRIPTION
**Pull request by shiltian**
`-parallel-jobs` is not always available, such as upstream LLVM.

## Details
___Do not mention proprietary info or link to internal work items in this PR.___

**Work item:** _"Internal", or link to GitHub issue (if applicable)._

**What were the changes?**  
_One sentence describing the work done._

**Why were the changes made?**  
_Explain the motivation behind the work. Provide any publicly-available historical context._

**How was the outcome achieved?**  
_Technical details behind the work. Explain any publicly-available hardware peculiarities._

**Additional Documentation:**  
_What else should the reviewer know?_

## Approval Checklist
___Do not approve until these items are satisfied.___
- [ ] Verify the CHANGELOG has been updated, if
  - there are any NCCL API version changes,
  - any changes impact library users, and/or
  - any changes impact any other ROCm library.

### CODE REVIEW

**Comment by wenkaidu** [comment]
@shiltian how to do parallel compilation and linking without -parallel-jobs?

**Comment by shiltian** [comment]
Parallel compilation is typically managed by the build system, such as `make` or `ninja`, using options like `-j N`. There's no concept of parallel linking outside of scenarios like LTO/ThinLTO. In RDC mode, LTO is the default, but for our current pipeline, the parallel part isn't particularly critical.

The `-parallel-jobs` flag, however, is specific to ROCm Clang and is not supported by upstream LLVM. Since I often use upstream LLVM to compile RCCL, it’s inconvenient to manually comment out the `-parallel-jobs` flag every time.

This PR adds the flag only if the compiler supports it, avoiding unnecessary compile errors without changing current behavior.

On a side note, I'm currently trying to enable ThinLTO for AMDGPU and using RCCL for my experiment. That's how I found out those CMake improvements.

**Review by wenkaidu — APPROVED** [review]

File: CMakeLists.txt [hunk 1]
```diff
@@ -805,12 +812,14 @@ else()
   endif()
 endif()
 ## Reserve 16GB for each linker job. Limit max number of linker jobs to 16
-math(EXPR num_linker_jobs "(${memory_in_gb} + 15) / 16")
-if (${num_linker_jobs} GREATER_EQUAL "16")
-  set(num_linker_jobs "16")
+if(HAVE_PARALLEL_JOBS)
```
Comment (shiltian) [inline CMakeLists.txt]: I figure there should be a white space.

**Comment by shiltian** [comment]
I don't have merge permit. Can anyone with permit help with it? Thanks.

**Pull request merged**